game-core/src/main/java/games/strategy/triplea/ui/PlacementUnitsCollapsiblePanel.java

```diff
@@ -0,0 +1,76 @@
+package games.strategy.triplea.ui;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.GameDataEvent;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.GameStep;
+import games.strategy.triplea.util.UnitSeparator;
+import javax.swing.SwingUtilities;
+import lombok.Getter;
+import org.triplea.swing.CollapsiblePanel;
+
+class PlacementUnitsCollapsiblePanel {
+  private final SimpleUnitPanel unitsToPlacePanel;
+
+  @Getter private final CollapsiblePanel panel;
+  private final GameData gameData;
+
+  PlacementUnitsCollapsiblePanel(final GameData gameData, final UiContext uiContext) {
+    this.gameData = gameData;
+    unitsToPlacePanel =
+        new SimpleUnitPanel(
+            uiContext, SimpleUnitPanel.Style.SMALL_ICONS_WRAPPED_WITH_LABEL_WHEN_EMPTY);
+    panel = new CollapsiblePanel(unitsToPlacePanel, "");
+    panel.setVisible(false);
+    gameData.addGameDataEventListener(GameDataEvent.GAME_STEP_CHANGED, this::updateStep);
+  }
+
+  private void updateStep() {
+    final GameStep step = gameData.getSequence().getStep();
+
+    SwingUtilities.invokeLater(
+        () -> {
+          if (GameStep.isPlaceStep(step.getName()) || isInitializationStep(step)) {
+            panel.setVisible(false);
+            return;
+          }
+
+          final boolean hasUnitsToPlace = !step.getPlayerId().getUnits().isEmpty();
+
+          if (hasUnitsToPlace || stepIsAfterPurchaseAndBeforePlacement(step)) {
+            panel.setTitle("Units To Place (" + step.getPlayerId().getUnits().size() + ")");
+            unitsToPlacePanel.setUnitsFromCategories(
+                UnitSeparator.categorize(step.getPlayerId().getUnits()));
+            panel.setVisible(true);
+          } else {
+            panel.setVisible(false);
+          }
+        });
+  }
+
+  private static boolean isInitializationStep(final GameStep gameStep) {
+    return gameStep.getPlayerId() == null;
+  }
+
+  private boolean stepIsAfterPurchaseAndBeforePlacement(final GameStep step) {
+    final GamePlayer currentPlayer = step.getPlayerId();
+
+    for (int i = gameData.getSequence().getStepIndex() - 1; i >= 0; i--) {
```

Comment:
Is there any specific reason why we iterate backwards over the steps?

Reply:
Yes, we're trying to see if given a current step, that within the same turn if a player has passed through a purchase phase. Some maps have a combat move before purchase, so we can't just check if we are in a combat move.

I'll note as well, there's still an assumption that there will be only one place phase, that it has a specific name too, all of which is already a bit brittle.

Reply:
Ok, but why is the order important then?
Wouldn't it be sufficient to iterate from 0 to step index over the elements?

Reply:
Perhaps important to keep in mind that steps is all steps across a full round of play, it has a length of around 70. If we go from 0, we will iterate more, but more importantly it's harder to know when we've dropped to the previously players turn.